### PR TITLE
Variations: Ask for user confirmation before creating variations

### LINKED ISSUE #8492: Variations: Inform about variations to be created
We need to notify the user how many variations will be created before actually creating the variations.

Image

Also notify when there are no variations to be created

### PR DESCRIPTION
**Pull request by Ecarrion**
Closes: #8492 

# Why

This PR asks the merchant for confirmation before generating multiple variations remotely.

# How

- Refactor `ViewModel.generateAllVariations` method to provide an `onStateChanged` closure that will inform the consumer what is happening in the generation process.

- Update View controller to listen to those state changes.

- Presents confirmation alert when the state change indicates. 

# Demo

https://user-images.githubusercontent.com/562080/210891714-bfac3759-deff-4a6a-a3e1-2c1a1a10ec28.mov

# Testing Step

- Go to a variable product
- Add some attributes and options to the product
- Start the "Generate All Variations" flow
- See that an alert is presented to the merchant to confirm the generation process.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Products/Variations/ProductVariationsViewController.swift [hunk 1]
```diff
@@ -820,6 +842,17 @@ private extension ProductVariationsViewController {
...
                                                               comment: "Error title when failing to generate a variation.")
         static let variationCreated = NSLocalizedString("Variation created", comment: "Text for the notice after creating the first variation.")
+
+        static let confirmationTitle = NSLocalizedString("Generate all variations?",
+                                                         comment: "Alert title to allow the user confirm if they want to generate all variations")
+        static func confirmationDescription(variationCount: Int) -> String {
+            let format = NSLocalizedString("This will create a variation for each and every possible combination of variation attributes (%d variations).",
+                                           comment: "Alert description to allow the user confirm if they want to generate all variations")
+            return String.localizedStringWithFormat(format, variationCount)
+        }
```
Comment (ealeksandrov) [inline WooCommerce/Classes/ViewRelated/Products/Variations/ProductVariationsViewController.swift:852]: Nice wrapper!

File: WooCommerce/Classes/ViewRelated/Products/Variations/ProductVariationsViewModel.swift [hunk 2]
```diff
@@ -26,27 +26,49 @@ final class ProductVariationsViewModel {
...
+                // Guard for no variations to generate
                 guard variationsToGenerate.count > 0 else {
-                    // TODO: Inform user that no variation will be created
-                    return onCompletion(.success(()))
+                    return onStateChanged(.finished(false))
                 }
 
-                self.createVariationsRemotely(for: product, variations: variationsToGenerate, onCompletion: onCompletion)
+                // Confirm generation with merchant
+                onStateChanged(.confirmation(variationsToGenerate.count, { confirmed in
```
Comment (ealeksandrov) [inline WooCommerce/Classes/ViewRelated/Products/Variations/ProductVariationsViewModel.swift:55]: This one is hard to read and understand, but I don't have easy improvement suggestions 🤔

**Review by ealeksandrov — APPROVED** [review]
LGTM!

File: WooCommerce/Classes/ViewRelated/Products/Variations/ProductVariationsViewModel.swift (hunk 2 above)
Comment (Ecarrion) [reply WooCommerce/Classes/ViewRelated/Products/Variations/ProductVariationsViewModel.swift:55]: Yeah, not a fan either. will keep it in the back of my head for improvements!

**Pull request merged**